### PR TITLE
Implement map_overlap

### PR DESCRIPTION
**Pull request by phofl**
Planning on merging this soonish to unblock rolling work, but feedback welcome

### CODE REVIEW

File: dask_expr/_expr.py [hunk 1]
```diff
@@ -1275,6 +1276,196 @@ def _task(self, index: int):
             )
 
 
+class MapOverlap(MapPartitions):
```
Comment (rjzamora) [inline dask_expr/_expr.py:1279]: Should we keep this in a new `dask_expr/_rolling.py`?

File: dask_expr/tests/test_collection.py [hunk 2]
```diff
@@ -1291,3 +1291,59 @@ def test_replace_filtered_combine_similar():
...
+    result = df.map_overlap(func, before=0, after="2D")
+    expected = lib.DataFrame([5, 5, 5, 3, 3], index=idx, columns=["a"])
+    assert_eq(result, expected)
+    result = df.map_overlap(func, before=0, after=1)
+    assert_eq(result, expected)
+
+    # Bug in dask/dask
+    # result = df.map_overlap(func, before=0, after="1D")
+    # expected = lib.DataFrame([4, 4, 4, 3, 3], index=idx, columns=["a"])
+    # assert_eq(result, expected)
```
Comment (rjzamora) [inline dask_expr/tests/test_collection.py:1314]: Is there an issue open?

**Review by rjzamora — COMMENTED** [review]
Thanks for putting this together.  This PR is relatively large, and you are signaling that it is also time sensitive. Are there specific areas/concerns that you are unsure of or are interested in feedback on?

File: dask_expr/tests/test_collection.py (hunk 2 above)
Comment (phofl) [reply dask_expr/tests/test_collection.py:1314]: https://github.com/dask/dask/pull/10639

**Comment by phofl** [comment]
The complexity is in class MapOverlapInterleavePartitions(Expr):, everything else is fairly straightforward, so a set of eyeballs would be appreciated there

File: dask_expr/_expr.py (hunk 1 above)
Comment (phofl) [reply dask_expr/_expr.py:1279]: Hm I'd rather not, this is not rolling specific (fillna uses this as well for example)

Comment (rjzamora) [reply dask_expr/_expr.py:1279]: Okay cool. I noticed that the `dask.dataframe` version was in a `rolling` module, so I just wanted to check.

Comment (phofl) [reply dask_expr/_expr.py:1279]: Yeah that part confused me :)

Comment (rjzamora) [inline dask_expr/_expr.py:1279]: Mostly a "note to self": I'm wondering if there is any "danger" in treating the pre-lowered version of `MapOverlap` as a `Blockwise` expression. It is very important that we don't do any partition-related optimizations (e.g. culling) until after this expression is lowered. It seems like we are in the clear for the current `optimize` implementation, but it isn't clear to me that we are gaining much by inheriting from `MapPartitions`  to begin with?

File: dask_expr/_expr.py [hunk 3]
```diff
@@ -1275,6 +1276,196 @@ def _task(self, index: int):
...
+            _overlap_chunk,
+            self._meta,
+            self.enforce_metadata,
+            self.transform_divisions,
+            self.clear_divisions,
+            self._kwargs,
+        )
+
+
+class MapOverlapInterleavePartitions(Expr):
```
Comment (rjzamora) [inline dask_expr/_expr.py:1348]: Another note to self: The `Interleave` term comes from the fact that we are extracting before/after rows for each partition and effectively "interleaving" the duplicated data between the "real" partitions. I don't love this name, but I also don't really have a better suggestion (maybe `CreateOverlappingPartitions`?)

**Review by rjzamora — COMMENTED** [review]
Still working through this whenever I have free moments. Not seeing any serious problems yet.

File: dask_expr/_expr.py (hunk 1 above)
Comment (phofl) [reply dask_expr/_expr.py:1279]: Culling in general has to happen after lower, but I agree that this isn't ideal.

Maybe we have to create a new class that sits somewhere in between Expr and Blockwise, but removing this information removes a lot of information that is helpful

File: dask_expr/_expr.py (hunk 3 above)
Comment (phofl) [reply dask_expr/_expr.py:1348]: No strong opinion either way

**Comment by phofl** [comment]
I am merging this to unblock rolling work, but happy to iterate further

**Pull request merged**